refactor: acquireAsync will dispatch a job, not be dispatched

ximinez opened the pull request:
## High Level Overview of Change

Does a few different things:
1. Adds a new RAII class `CanProcess` to use where needed for job deduplication.
2. Adds and improves logging across a variety of classes. Uses the new `to_short_string(base_uint)` function to help disambiguate some messages about peers and ledgers.
3. Exits `InboundLedger::onTimer` without sending new requests if the complete ledger is available. (Duh.)
4. Changes the interface and usage of `InboundLedgers::acquireAsync` so that it dispatches a job, and is not called from a dispatched job. This will reduce the number of jobs spun up only to immediately exit.

### Context of Change

This is the safe subset of changes from #5301, which has problems, plus the item to dispatch from `acquireAsync`.

### Type of Change

- [X] Refactor (non-breaking change that only restructures code)
- [X] Performance (increase or change in throughput and/or latency)
- [X] Tests (you added tests for code that already exists, or your new feature included in this PR)

### API Impact

None

ximinez commented:
Unfortunately, this is not ready for prime time. I ran a simple test over the weekend, running this branch for some time, and running `develop` for some time.

What the data shows is:
* This branch ran for about 14 hours. In that time, it went out of sync 239 times. (`239 NetworkOPs:WRN STATE->connected - updateOperatingMode: no positions`), which is about 17 de-syncs per hour.
* `develop` ran for about 48 hours. In that time, it never dropped all the way down to "connected", and it went to "syncing" 43 times. At least of those had to be during initial sync, so call it 42. Either way, that's < 1 de-sync per hour.

Clearly, something is misbehaving, so I introduced a problem somewhere. I may be able to take some time to find out where, but it's not my priority right now, so it may be a while.

```
$ for f in acquireAsync.txt upstream.txt ; do echo -e "\n$f" ; head -1 $f ; tail -1 $f ; loganalysis $f; done

acquireAsync.txt
2025-Apr-25 22:41:05.433965087 UTC Application:NFO Process starting: rippled-2.4.0+7d8411985cc62926834ac06198422cb101feca0e.DEBUG, Instance Cookie: 17903171231208261479
2025-Apr-26 12:32:38.881102671 UTC NetworkOPs:NFO STATE->full - endConsensus: check full
      4 appeared;  following new file
      1 Application:NFO Process starting: rippled-#+7d8411985cc62926834ac06198422cb101feca0e.DEBUG, Instance Cookie: #
      4 become inaccessible: No such file or directory
      1 NetworkOPs:NFO STATE->connected - Heartbeat: sufficient peers
    291 NetworkOPs:NFO STATE->full - endConsensus: check full
      1 NetworkOPs:NFO STATE->syncing - Heartbeat: check connected
    296 NetworkOPs:NFO STATE->tracking - endConsensus: check tracking
      1 NetworkOPs:WRN STATE->connected - Heartbeat: check syncing
    239 NetworkOPs:WRN STATE->connected - updateOperatingMode: no positions
      2 NetworkOPs:WRN STATE->syncing - check LCL: not on consensus ledger
     25 NetworkOPs:WRN STATE->syncing - consensusViewChange
     29 NetworkOPs:WRN STATE->syncing - updateOperatingMode: no positions

upstream.txt
2025-Apr-26 14:11:56.660217534 UTC Application:NFO Process starting: rippled-2.4.0+fa1e25abef7bf12bde87ba23528346853cef88b0.DEBUG, Instance Cookie: 7137182022843094303
2025-Apr-28 14:21:47.931018040 UTC NetworkOPs:NFO STATE->full
     16 appeared;  following new file
      1 Application:NFO Process starting: rippled-#+fa1e25abef7bf12bde87ba23528346853cef88b0.DEBUG, Instance Cookie: #
     16 become inaccessible: No such file or directory
      1 NetworkOPs:NFO STATE->connected
     40 NetworkOPs:NFO STATE->full
     43 NetworkOPs:NFO STATE->syncing
     43 NetworkOPs:NFO STATE->tracking
```

ximinez commented:
This is ready for review. I accidentally ran the original code in the test mentioned in my previous comment. When I ran it again over the last few days, it had a more de-syncs, but not unreasonably so. About 40 in the last 24 hours. Maybe still enough to be a problem, but not the disaster I observed earlier.
```
     38 NetworkOPs:NFO STATE->full - endConsensus: check full
     39 NetworkOPs:NFO STATE->tracking - endConsensus: check tracking
      4 NetworkOPs:WRN STATE->connected - updateOperatingMode: no positions
      4 NetworkOPs:WRN STATE->syncing - check LCL: not on consensus ledger
     27 NetworkOPs:WRN STATE->syncing - consensusViewChange
      4 NetworkOPs:WRN STATE->syncing - updateOperatingMode: no positions
```

vvysokikh1 reviewed (commented):
I'm a bit conflicted about this change. I think it should work, but I don't like the design. Would it be better to implement wrapper around container type that would provide such properties instead of RAII wrapper around iterator?

ximinez committed:
refactor: acquireAsync will dispatch a job, not the other way around

- Improve job queue collision checks and logging
    - Improve logging related to ledger acquisition and operating mode
      changes
    - Class "CanProcess" to keep track of processing of distinct items

ximinez commented:
> [ximinez](https://github.com/ximinez) [force-pushed](https://github.com/XRPLF/rippled/compare/877549e57aa18123504813c039e09f1394703bac..f44dfc89cca024ff823dee56892c8c9f555f8b4f) the ximinez/acquireAsyncDispatch branch from [877549e](https://github.com/XRPLF/rippled/commit/877549e57aa18123504813c039e09f1394703bac) to [f44dfc8](https://github.com/XRPLF/rippled/commit/f44dfc89cca024ff823dee56892c8c9f555f8b4f)

I squashed and rebased onto current develop to resolve the missing commit signatures issue without needing to re-resolve a bunch of merge conflicts.

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/acquireAsyncDispatch

* XRPLF/develop:
  chore: Enable clang-tidy include cleaner (6947)
  fix: Change AMMClawback return code to tecNO_PERMISSION (6946)
  ci: [DEPENDABOT] bump actions/upload-pages-artifact from 4.0.0 to 5.0.0 (6927)
  ci: [DEPENDABOT] bump actions/upload-artifact from 7.0.0 to 7.0.1 (6928)
  chore: Enable clang-tidy readability checks (6930)

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/acquireAsyncDispatch

* XRPLF/develop:
  chore: Add -fix to clang-tidy invocation (6990)
  chore: Remove empty Taker.h (6984)
  chore: Enable clang-tidy modernize checks (6975)
  ci: Upload clang-tidy git diff (6983)
  fix: Add rounding to Vault invariants (6217) (6955)

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/acquireAsyncDispatch

* XRPLF/develop:
  ci: Rewrite clang-tidy workflow(s) in a reusable manner (7062)
  chore: Ignore identifier-naming update in git blame (7066)
  refactor: Enable clang-tidy `readability-identifier-naming` check (6571)

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Fix new header to use "#pragma once"

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch

ximinez committed:
Merge branch 'develop' into ximinez/acquireAsyncDispatch
